### ISSUE TITLE
Support dependsOn depending on all chart resources

### ISSUE THREAD

**Issue by nesl247** [pulumi/pulumi-kubernetes#861]
### Problem description
Currently `dependsOn` only waits until Pulumi registers the chart, not until it's fully available.

See: https://github.com/pulumi/pulumi/issues/1743 and https://github.com/pulumi/pulumi-kubernetes/issues/814 (hack)

**Comment by pgavlin** [comment]
This is a consequence of https://github.com/pulumi/pulumi/issues/3282.

**Comment by 0x53A** [comment]
this seems to be similar for ``k8s.yaml.ConfigGroup``:

```ts

// Set up TLS
// -------------------------------------------------------------------------------------------------------------------------------------------------------

const certManagerCustomResources = new k8s.yaml.ConfigGroup(`${prefix}-k8-tls-custom-resources`, {
    files: ["https://raw.githubusercontent.com/jetstack/cert-manager/release-0.9/deploy/manifests/00-crds.yaml"]
}, { provider: k8sProvider });



const certmanagerNamespace = new k8s.core.v1.Namespace(`${prefix}-k8-certmanager-namespace`, {
    metadata: {
        name: "cert-manager",
        labels: { "certmanager.k8s.io/disable-validation": "true" }
    }
  }, { provider: k8sProvider, dependsOn: [certManagerCustomResources] });



const certManagerDeployment = new k8s.helm.v2.Chart(`cert-manager`, {
    repo: "jetstack",
    chart: "cert-manager",
    namespace: certmanagerNamespace.metadata.name,
    version: "v0.9.0",
    values: {
        nodeSelector: { "beta.kubernetes.io/os": "linux" },
        cainjector: { nodeSelector: { "beta.kubernetes.io/os": "linux" } },
        webhook: { nodeSelector: { "beta.kubernetes.io/os": "linux" } },
    }
}, {
    providers: { kubernetes: k8sProvider },
    dependsOn: [ certmanagerNamespace, certManagerCustomResources ]
});


const clusterIssuerYaml = `apiVersion: certmanager.k8s.io/v1alpha1
kind: ClusterIssuer
metadata:
  name: ${letsencryptName}
  namespace: ingress-basic
spec:
  acme:
    server: ${acmeServer}
    email: ${letsEncryptEmail}
    privateKeySecretRef:
      name: ${letsencryptName}
    http01: {}
    solvers:
    - http01:
        ingress:
          podTemplate:
            spec:
              nodeSelector:
                "beta.kubernetes.io/os": "linux"`;


const clusterIssuerDeployment = new k8s.yaml.ConfigGroup(`${prefix}-k8-certmanager-cluster-issuer`, { yaml: [ clusterIssuerYaml ]}, { provider: k8sProvider, dependsOn: [ certManagerDeployment ] });

```

The namespace depends on the CRD ConfigGroup, and everything else obviously depends on the namespace, so it should only continue after the CRDs have been created.

![image](https://user-images.githubusercontent.com/4236651/68145719-b31f4a80-ff36-11e9-9526-fd399f5dd59a.png)

As you can see it tried to create the cluster issuer before the CRDs have been created.

This makes the whole thing really fragile.

**Comment by ulrichSchreiner** [comment]
hi,
any work on this? i have exactly the same problem with the cert-manager manifests and dependent resssources.

**Comment by mitchellmaler** [comment]
I think I just ran into this issue as well using .NET. I need to deploy the rancher helm release before I create a rancher2.bootstrap resource. I have the bootstrap resource depending on the helm release but it seems to be ignored due to it always trying to run first which ends with an error since rancher isn't installed yet.

```c#
var rancherHelmRelease = new Pulumi.Kubernetes.Helm.V3.Chart("rancher",
    new ChartArgs
    {
        Repo = "rancher-latest",
        Chart = "rancher",
        Version = config.Require("rancherVersion"),
        Namespace = cattleSystemNamespace.Metadata.Apply(x => x.Name),
        Values =
        {
            {"ingress.tls.source", "external"},
            {"tls", "external"},
            {"hostname", config.Require("rancherHost")},
            {"debug", config.RequireBoolean("rancherEnableDebug")},
            {"replicas", 3}
        }
    },
    new ComponentResourceOptions
    {
        Provider = kubernetesProvider,
        DependsOn =
        {
            rkeCluster,
            cattleSystemNamespace
        }
    });
var rancherBootstrapProvider = new Pulumi.Rancher2.Provider("rancherBootstrapProvider",
    new Pulumi.Rancher2.ProviderArgs
    {
        ApiUrl = $"https://{config.Require("rancherHost")}",
        Bootstrap = true
    });
// Initially setup server the first time creating admin account
var rancherBootstrap = new Pulumi.Rancher2.Bootstrap("rancherBootstrap",
    new BootstrapArgs
    {
        Password = commonEnvSecret.Data.Apply(x => (string)x["rancher_server_admin_password"]),
        Telemetry = false
    },
    new CustomResourceOptions
    {
        Provider = rancherBootstrapProvider,
        DependsOn =
        {
            rancherHelmRelease
        }
    });
```

**Comment by mitchellmaler** [comment]
So how can I return the list of resources or even the resource from the component so that another resource can depend on it? It does not let me return Outputs that are not "String, Boolean, Int32, Double, Nullable<...>, ImmutableArray<...> and ImmutableDictionary<string, ...> or a class explicitly marked with the [OutputTypeAttribute]" from ComponentResource classes.

The Helm resource is also not customizable since it is a library so how can I get the list of resources it creates as a property so I can add it for depends on?

**Comment by mitchellmaler** [comment]
Was able to solve the issue with the components to creating a class property that holds a reference to all the created resources.

@mikhailshilkov Can a new property be created on the Pulumi.Kubernetes.Helm.V3.Chart class that tracks all resources and returns them? Looks like CollectionComponentResource has a Resources property that tracks it and a method GetResource that returns a single resource if you know what you are looking for. Would be great if we could reference the resources property or a new method to returns all resources.

**Comment by mikhailshilkov** [comment]
Sounds reasonable. @lblackstone have we considered exposing child resources from a Helm chart?

**Comment by lblackstone** [comment]
> have we considered exposing child resources from a Helm chart?

I haven't tried that, but it sounds like it could be a reasonable solution here. I'm open to trying it.

If that works out, it would be worth seeing if we could make that solution more general in the `ComponentResource` class.

**Comment by jmgilman** [comment]
I'm slightly confused on how one would depend on a resource created by the chart. I would of thought something like this would have worked:

```
crd = self.chart.get_resource(group_version_kind='apiextensions.k8s.io/v1/CustomResourceDefinition',
                                      name='clusterissuers.cert-manager.io',
                                      namespace=self.props.namespace_str).apply(lambda res: res)
```
...
opts=pulumi.ResourceOptions(parent=self, depends_on=[crd, self.token_secret]))
...
```

This results in:

```
Exception: 'depends_on' was passed a value that was not a Resource.
```

The [documentation](https://www.pulumi.com/docs/reference/pkg/python/pulumi_kubernetes/helm/v3/#pulumi_kubernetes.helm.v3.Chart.get_resource) makes it seem like if I unwrapped the result from `get_resource` I would get a `CustomResource` I could then depend on. Am I doing something wrong here?

**Comment by manast** [comment]
@jmgilman I am suffering your pain too. Seems like a ticking bomb, using some resource whose dependencies will not be correctly tracked, has high probability to break not only the deployment but in my experience the stack as well, which in production is really a no-go. 
If we could at least get a list of which resources not to use so that we can avoid them completely, otherwise this is like a recipe for disaster sooner or later...

**Comment by mitchellmaler** [comment]
@lblackstone Is there any workaround for this issue? We deploy many helm charts and it is causing a lot of headaches.
I think just having a list of Resources exported from the Chart release would fix this issue. We could then depend on that list of a single item from that list.

**Comment by lblackstone** [comment]
@mitchellmaler I've opened #1364 to investigate this -- that approach seems to be working for TypeScript, but we need to decide on the best way to do this, and make sure it works for all languages.

**Comment by mitchellmaler** [comment]
@lblackstone Great news. Really like the idea of being able to wait for the whole chart to be created.

**Comment by lblackstone** [comment]
Although TypeScript is working now, the other languages will require changes in the Pulumi SDK to support (the dependsOn opt needs to support Inputs rather than just the raw Resource). Since this might be a breaking change, it's going to require more investigation/discussion.

https://github.com/pulumi/pulumi/issues/5642 is related.

**Comment by harryttd** [comment]
I'm not sure if this is related exactly to this issue:

We are using gh actions to  run pulumi up. We have local Helm charts on our computers. One is a dep of the other. We reference the charts in the `path` prop when instantiating a new chart via `new k8s.helm.v2.Chart`.  We tried deploying both charts, but pulumi fails with:
```
  +  pulumi:pulumi:Stack tqinfra-dev creating Error: found in Chart.yaml, but missing in charts/ directory: authenticated-rpc-tezos-chain
   +  pulumi:pulumi:Stack tqinfra-dev creating Error: Error: Command failed: helm template mainnet_bcd/tezos-k8s/charts/tezos --name-template bcd-main --values mainnet_bcd/tezos-k8s/charts/tezos/values.yaml --values /tmp/tmp-2059gBUNjUvO4U4E.yaml  --namespace tezos --include-crds
   +  pulumi:pulumi:Stack tqinfra-dev creating Error: found in Chart.yaml, but missing in charts/ directory: authenticated-rpc-tezos-chain
   +  pulumi:pulumi:Stack tqinfra-dev creating     at /home/runner/work/tqinfra/tqinfra/node_modules/@pulumi/helm/v2/helm.ts:250:23
   +  pulumi:pulumi:Stack tqinfra-dev creating     at /home/runner/work/tqinfra/tqinfra/node_modules/@pulumi/pulumi/output.js:249:35
   +  pulumi:pulumi:Stack tqinfra-dev creating     at Generator.next (<anonymous>)
   +  pulumi:pulumi:Stack tqinfra-dev creating     at /home/runner/work/tqinfra/tqinfra/node_modules/@pulumi/pulumi/output.js:21:71
   +  pulumi:pulumi:Stack tqinfra-dev creating     at new Promise (<anonymous>)
   +  pulumi:pulumi:Stack tqinfra-dev creating     at __awaiter (/home/runner/work/tqinfra/tqinfra/node_modules/@pulumi/pulumi/output.js:17:12)
   +  pulumi:pulumi:Stack tqinfra-dev creating     at applyHelperAsync (/home/runner/work/tqinfra/tqinfra/node_modules/@pulumi/pulumi/output.js:228:12)
   +  pulumi:pulumi:Stack tqinfra-dev creating     at /home/runner/work/tqinfra/tqinfra/node_modules/@pulumi/pulumi/output.js:182:65
   +  pulumi:pulumi:Stack tqinfra-dev creating     at processTicksAndRejections (internal/process/task_queues.js:93:5)
   ```

The problem may be that pulumi is not running `helm dependency update` on the parent chart.

**Comment by gitfool** [comment]
This is a nightmare. I can't control order of creation (and deletion) of an ever increasing web of helm charts and yaml config groups. I shouldn't have to run a deployment 3 or more times to work around all the races due to lack of recursive dependency tracking. Not sure how to proceed unless I'm missing a workaround for dotnet/c#. I feel like I'm hosed until this is fixed. 😞

**Comment by gitfool** [comment]
I was missing a workaround for dotnet/c# that uses [CollectionComponentResource.GetResource\<T\>](https://github.com/pulumi/pulumi-kubernetes/blob/241ddef1d8a9db3a507e536bd163c036cd98a8bd/sdk/dotnet/Yaml/Yaml.cs#L54) (from base class of Helm [Chart](https://github.com/pulumi/pulumi-kubernetes/blob/241ddef1d8a9db3a507e536bd163c036cd98a8bd/sdk/dotnet/Helm/V3/Chart.cs#L229)) to get some of the Helm chart resources so that I can explicitly depend on them to force order of create / delete. For example, cert manager is very slow to be ready, and it's also recommended to [run a test](https://github.com/jetstack/cert-manager/issues/2908#issuecomment-632751413) afterwards to be sure:
```c#
// cert manager; https://github.com/jetstack/cert-manager/tree/master/deploy/charts/cert-manager
Logger.LogDebug("Installing cert manager");
var certManagerCrds = new ConfigGroup("cert-manager-crds",
    new ConfigGroupArgs { Yaml = ReadResource("CertManagerCrds.yaml") },
    new ComponentResourceOptions { /*Protect = true,*/ Provider = k8sProvider });

var certManagerNs = new Namespace("cert-manager",
    new NamespaceArgs { Metadata = new ObjectMetaArgs { Name = "cert-manager" } },
    new CustomResourceOptions { Provider = k8sProvider });

var certManagerValues =
    new Dictionary<string, object>
    {
        ["prometheus"] = new
        {
            enabled = true,
            servicemonitor = new { enabled = true }
        }
    };

var certManagerChart = new Chart("cert-manager",
    new ChartArgs
    {
        Namespace = "cert-manager",
        FetchOptions = new ChartFetchArgs { Repo = "https://charts.jetstack.io" },
        Chart = "cert-manager",
        Version = K8sConfig.CertManagerChartVersion,
        Values = certManagerValues,
        SkipCRDRendering = true
    },
    new ComponentResourceOptions { DependsOn = { certManagerCrds, certManagerNs, kubePrometheusStackCrds }, Provider = k8sProvider });

var certManager = certManagerChart.GetResource<Deployment>("cert-manager", "cert-manager").AsResource();
var certManagerCaInjector = certManagerChart.GetResource<Deployment>("cert-manager-cainjector", "cert-manager").AsResource();
var certManagerWebHook = certManagerChart.GetResource<Deployment>("cert-manager-webhook", "cert-manager").AsResource();

var certManagerTest = new ConfigGroup("cert-manager-test",
    new ConfigGroupArgs { Yaml = RenderTemplate("CertManagerTest.yaml", ReadResource, new { @namespace = "cert-manager" }) },
    new ComponentResourceOptions { DependsOn = { certManager, certManagerCaInjector, certManagerWebHook }, Provider = k8sProvider });

var certManagerTestCert = certManagerTest.GetCustomResource("cert-manager.io/v1/Certificate", "cert-manager-test-cert", "cert-manager").AsResource();
//var certManagerTestIssuer = certManagerTest.GetCustomResource("cert-manager.io/v1/Issuer", "cert-manager-test-issuer", "cert-manager").AsResource();
```
Any helm chart that uses cert manager can then explicitly depend on `certManagerTestCert` to avoid any race conditions.

I think this could be improved significantly by adding another method that returns all the internal [resources](https://github.com/pulumi/pulumi-kubernetes/blob/241ddef1d8a9db3a507e536bd163c036cd98a8bd/sdk/dotnet/Yaml/Yaml.cs#L35) for the helm chart or yaml config group as an array with something like:
```c#
public abstract class CollectionComponentResource : ComponentResource
{
    public Output<ImmutableArray<Resource>> GetResources
        => Resources.Apply(resources => resources.Values.Cast<Resource>().ToImmutableArray());
}
```
That would tide us over until it becomes the default behavior for helm charts and yaml config groups.

**Comment by stanislav-zaprudskiy** [comment]
This is what worked for me:

```python
cert_manager = Chart(
    "cert-manager",
    ChartOpts(
        chart="cert-manager",
        version=f'v{CERT_MANAGER_VERSION}',
        fetch_opts=FetchOpts(repo="https://charts.jetstack.io"),
        namespace=cert_manager_namespace.metadata.name,
        values={
            "installCRDs": "true"
        },
        transformations=[remove_status],
    ),
    opts=ResourceOptions(provider=k8s_provider),
)

actions_controller = Chart(
    "actions-runner-controller",
    ChartOpts(
        chart="actions-runner-controller",
        version="0.12.7",
        fetch_opts=FetchOpts(
            repo="https://actions-runner-controller.github.io/actions-runner-controller"
        ),
        namespace=actions_runner_controller_namespace.metadata.name,
        transformations=[remove_status],
    ),
    opts=ResourceOptions(
        provider=k8s_provider,
        depends_on=cert_manager.resources.apply(lambda resources: list(resources.values())),
    ),
)
```

**Comment by lblackstone** [comment]
Alright, I'm pleased to announce that we finally have this fixed for all SDKs in the [v3.9.0](https://github.com/pulumi/pulumi-kubernetes/releases/tag/v3.9.0) release! In each case, you can depend on a Chart by providing the `ready` property to the `dependsOn` resource option, and the dependent resource will wait until all of the chart resources are ready before continuing. 

(Depending on the `ready` property rather than the resource object directly was necessary because the implementation relies on resources created inside an apply, which is [generally not recommended](https://www.pulumi.com/docs/intro/concepts/inputs-outputs/#apply).)

I'm still working on adding an example to the API docs, but I'll link to the tests that show the usage in the meantime:

[NodeJS](https://github.com/pulumi/pulumi-kubernetes/blob/b6e74c9613ef3b4c9492e41417b529b592a99abc/tests/sdk/nodejs/examples/helm-local/step1/index.ts#L44-L47)
[Python](https://github.com/pulumi/pulumi-kubernetes/blob/b6e74c9613ef3b4c9492e41417b529b592a99abc/tests/sdk/python/helm-local/step1/__main__.py#L21-L22)
[Go](https://github.com/pulumi/pulumi-kubernetes/blob/b6e74c9613ef3b4c9492e41417b529b592a99abc/tests/sdk/go/helm-local/step1/main.go#L27-L31)
[.NET](https://github.com/pulumi/pulumi-kubernetes/blob/b6e74c9613ef3b4c9492e41417b529b592a99abc/tests/sdk/dotnet/helm-local/step1/Program.cs#L19-L28)

**Comment by lblackstone** [comment]
Alright, I [added an example](https://www.pulumi.com/registry/packages/kubernetes/api-docs/helm/v3/chart/#depend-on-a-chart-resource) to the API Reference docs, so I'm finally closing this out!

**Comment by lblackstone** [comment]
> Are there any other cases where non-standard workarounds are considered the solution? I ask this because using `.ready` is different than any other `dependsOn` from my knowledge. Shouldn't pulumi only consider the ticket closed when it behaves as all other resources? Anyone expecting this will not know to look at one specific example that is different than every other resource.

Helm Chart is the only one so far, but I'd consider it a stopgap solution. Eventually, we'd like to make it possible to do a regular `dependsOn` for any ComponentResource that works as you'd expect. This work is ongoing, but doesn't have a definitive ETA yet, so we decided on the `Ready` approach to unblock users in the meantime. I realize that this may still trip up new users, but it seems like the most pragmatic solution since this has been a commonly encountered problem for a long time.

I'm certainly open to suggestions for improving the documentation, but I think this solution solves the core issue, so I'd prefer to track any additional work in separate issues.

**Comment by gitfool** [comment]
@lblackstone this has been a [painful trap](https://github.com/pulumi/pulumi/issues/7422#issuecomment-978351763) with `ConfigGroup` too.

**Comment by lblackstone** [comment]
@gitfool Yep, understood. I didn't see another issue specifically tracking ConfigFile/ConfigGroup, so I opened https://github.com/pulumi/pulumi-kubernetes/issues/1833

**Comment by georgiosd** [comment]
I have the opposite problem! I am trying to execute a Helm Release (nginx ingress) AFTER the cluster is setup, but Pulumi tries to connect. `DependsOn = cluster` doesnt seem to do anything. Any ideas?

**Comment by keithharvey** [comment]
`ksqlServer.status.apply` solved the problem for me, with service lookup after a v3 helm release

**Comment by blaargh** [comment]
The `dependsOnInput(chart.Ready)` does not have the expected effect on Get calls. See example:

```
chart, err := helm.NewChart(ctx, "chartName", helm.ChartArgs{
      Chart:     pulumi.String("chart"),
      Namespace: pulumi.String("ns"),
}, nil, pulumi.Provider(k8sProvider))

if err != nil {
  log.Panic(err)
}

ingress, err := v1.GetIngress(ctx, "ingress", pulumi.ID("namespace/ingress"), nil, pulumi.DependsOnInputs(chart.Ready), pulumi.Provider(k8sProvider))
if err != nil {
  log.Panic(err)
}

hostname := ingress.Status.ApplyT(func(status *v1.IngressStatus) (string, error) {
    if len(status.LoadBalancer.Ingress) > 0 {
      	return *status.LoadBalancer.Ingress[0].Hostname, nil
   }
   return "", nil
}).(pulumi.StringOutput)
```

This errors during the preview with `error: Preview failed: resource 'namespace/ingress' does not exist`.

I see in the documentation it works with creating new resources, is it not supposed to work with a lookup? Or am I doing something wrong?

**Comment by mremes** [comment]
Any way to depend on multiple Helm charts?
